File: lib/features/profiles/domain/applicant_profile.dart
import '../../authentication/domain/user_account.dart';

enum EnglishTestStatus { notTaken, planned, completed, notRequired }

enum PassportStatus { unavailable, applied, valid, expired }

enum EmploymentStatus { student, employed, selfEmployed, unemployed, other }

class ProfileDocument {
  const ProfileDocument({
    required this.id,
    required this.name,
    required this.type,
    required this.uploadedAt,
  });

  final String id;
  final String name;
  final String type;
  final DateTime uploadedAt;
}

class ApplicantProfile {
  const ApplicantProfile({
    required this.userId,
    required this.fullName,
    required this.nationality,
    required this.countryOfResidence,
    required this.dateOfBirth,
    required this.gender,
    required this.highestQualification,
    required this.degreeField,
    required this.academicClassification,
    required this.graduationYear,
    required this.workExperienceYears,
    required this.preferredStudyLevels,
    required this.preferredCountries,
    required this.areasOfInterest,
    required this.englishTestStatus,
    required this.passportStatus,
    required this.employmentStatus,
    required this.fundingPreferences,
    required this.specialEligibilityCategories,
    required this.documents,
  });

  factory ApplicantProfile.empty(UserAccount account) => ApplicantProfile(
    userId: account.id,
    fullName: account.fullName,
    nationality: '',
    countryOfResidence: '',
    dateOfBirth: null,
    gender: '',
    highestQualification: '',
    degreeField: '',
    academicClassification: '',
    graduationYear: null,
    workExperienceYears: 0,
    preferredStudyLevels: const [],
    preferredCountries: const [],
    areasOfInterest: const [],
    englishTestStatus: EnglishTestStatus.notTaken,
    passportStatus: PassportStatus.unavailable,
    employmentStatus: EmploymentStatus.student,
    fundingPreferences: const [],
    specialEligibilityCategories: const [],
    documents: const [],
  );

  final String userId;
  final String fullName;
  final String nationality;
  final String countryOfResidence;
  final DateTime? dateOfBirth;
  final String gender;
  final String highestQualification;
  final String degreeField;
  final String academicClassification;
  final int? graduationYear;
  final double workExperienceYears;
  final List<String> preferredStudyLevels;
  final List<String> preferredCountries;
  final List<String> areasOfInterest;
  final EnglishTestStatus englishTestStatus;
  final PassportStatus passportStatus;
  final EmploymentStatus employmentStatus;
  final List<String> fundingPreferences;
  final List<String> specialEligibilityCategories;
  final List<ProfileDocument> documents;

  ApplicantProfile copyWith({List<String>? preferredCountries}) =>
      ApplicantProfile(
        userId: userId,
        fullName: fullName,
        nationality: nationality,
        countryOfResidence: countryOfResidence,
        dateOfBirth: dateOfBirth,
        gender: gender,
        highestQualification: highestQualification,
        degreeField: degreeField,
        academicClassification: academicClassification,
        graduationYear: graduationYear,
        workExperienceYears: workExperienceYears,
        preferredStudyLevels: preferredStudyLevels,
        preferredCountries: preferredCountries ?? this.preferredCountries,
        areasOfInterest: areasOfInterest,
        englishTestStatus: englishTestStatus,
        passportStatus: passportStatus,
        employmentStatus: employmentStatus,
        fundingPreferences: fundingPreferences,
        specialEligibilityCategories: specialEligibilityCategories,
        documents: documents,
      );
}

File: scholarsphere_backend/app/models/applicant_profile.py
import enum
from datetime import date, datetime

from sqlalchemy import Date, DateTime, Enum, Float, Integer, String, func
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base
from app.models.external_opportunity import JSONType


class EnglishTestStatus(str, enum.Enum):
    not_taken = "not_taken"
    planned = "planned"
    completed = "completed"
    not_required = "not_required"


class PassportStatus(str, enum.Enum):
    unavailable = "unavailable"
    applied = "applied"
    valid = "valid"
    expired = "expired"


class EmploymentStatus(str, enum.Enum):
    student = "student"
    employed = "employed"
    self_employed = "self_employed"
    unemployed = "unemployed"
    other = "other"


class ApplicantProfile(Base):
    """One row per applicant - a settings/profile table, not an event table,

    so the Firebase uid is the primary key itself, matching
    NotificationPreferences' identical shape.

    ``documents`` is deliberately not modeled here: the Dart domain's
    ``ProfileDocument`` is metadata-only (name/type/timestamp, no file
    path or URL), so there is nothing real to persist yet - real file
    storage belongs to a dedicated Documents feature, not bolted onto
    profile fields. The API always returns an empty documents list.
    """

    __tablename__ = "applicant_profiles"

    user_id: Mapped[str] = mapped_column(String(255), primary_key=True)
    full_name: Mapped[str] = mapped_column(String(255), nullable=False)
    nationality: Mapped[str] = mapped_column(String(255), default="", nullable=False)
    country_of_residence: Mapped[str] = mapped_column(String(255), default="", nullable=False)
    date_of_birth: Mapped[date | None] = mapped_column(Date)
    gender: Mapped[str] = mapped_column(String(64), default="", nullable=False)
    highest_qualification: Mapped[str] = mapped_column(String(255), default="", nullable=False)
    degree_field: Mapped[str] = mapped_column(String(255), default="", nullable=False)
    academic_classification: Mapped[str] = mapped_column(String(255), default="", nullable=False)
    graduation_year: Mapped[int | None] = mapped_column(Integer)
    work_experience_years: Mapped[float] = mapped_column(Float, default=0, nullable=False)
    preferred_study_levels: Mapped[list[str]] = mapped_column(JSONType, default=list, nullable=False)
    preferred_countries: Mapped[list[str]] = mapped_column(JSONType, default=list, nullable=False)
    areas_of_interest: Mapped[list[str]] = mapped_column(JSONType, default=list, nullable=False)
    english_test_status: Mapped[EnglishTestStatus] = mapped_column(
        Enum(EnglishTestStatus, native_enum=False),
        default=EnglishTestStatus.not_taken,
        nullable=False,
    )
    passport_status: Mapped[PassportStatus] = mapped_column(
        Enum(PassportStatus, native_enum=False),
        default=PassportStatus.unavailable,
        nullable=False,
    )
    employment_status: Mapped[EmploymentStatus] = mapped_column(
        Enum(EmploymentStatus, native_enum=False),
        default=EmploymentStatus.student,
        nullable=False,
    )
    funding_preferences: Mapped[list[str]] = mapped_column(JSONType, default=list, nullable=False)
    special_eligibility_categories: Mapped[list[str]] = mapped_column(
        JSONType, default=list, nullable=False
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )
